### DISCUSSION TITLE
AWS IoT WiFi - Compilation error

### DISCUSSION THREAD

**Discussion by AdelAlm** [earlephilhower/arduino-pico#2026] [General]
Hello,

I am currently working with Raspberry Pi Pico W with ATECC608 Secure Element using ArduinoECCX08 library.

My code is based on [Securely Connecting an Arduino MKR WiFi 1010 to AWS IoT Core](https://docs.arduino.cc/tutorials/mkr-wifi-1010/securely-connecting-an-arduino-mkr-wifi-1010-to-aws-iot-core/) as I want to be able to connect to MQTT broker using client certificate with its private key stored in the slot 0 of the Secure Element ATECC608.

Based on my research there is a problem between ArduinoBearSSL and BearSSL::* namespace functions because I am getting this error:
```bash
In file included from /home/XX/.arduino15/packages/rp2040/hardware/rp2040/3.7.2/libraries/WiFi/src/WiFiClientSecure.h:24,
                 from /home/XX/.arduino15/packages/rp2040/hardware/rp2040/3.7.2/libraries/WiFi/src/WiFi.h:9,
                 from /tmp/.arduinoIDE-unsaved2024126-480488-xb6owc.lqq3c/AWS_IoT_WiFi/AWS_IoT_WiFi.ino:26:
/home/XX/.arduino15/packages/rp2040/hardware/rp2040/3.7.2/libraries/WiFi/src/WiFiClientSecureBearSSL.h: In member function 'int BearSSL::WiFiClientSecureCtx::getMFLNStatus()':
/home/XX/.arduino15/packages/rp2040/hardware/rp2040/3.7.2/libraries/WiFi/src/WiFiClientSecureBearSSL.h:125:31: error: 'br_ssl_engine_get_mfln_negotiated' was not declared in this scope; did you mean 'br_ssl_engine_renegotiate'?
  125 |         return connected() && br_ssl_engine_get_mfln_negotiated(_eng);
      |                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                               br_ssl_engine_renegotiate
```

Any idea for solution? I think that I have to use ArduinoBearSSL because it defines the setEccSlot() function which is used to define the certificate private key slot.

Thanks

**Comment by earlephilhower** [comment]
That function is defined in the BearSSL headers which are included in the core, and there is an example that actually compiles in that function to show how to use MFLN (`BearSSL_MaxFragmentLength.ino`) that is built using CI.  So, looks like this can't be reproduced in the unmodified core (and I just successfully built the MFLN example myself locally).

I don't know exactly what you're doing, but if you're going to use an external encryption chip then you can't use the built-in WiFiClientSecure(BearSSL) and need to write your own using it as a base.  If you've changed the BearSSL backend library, that goes doubly-so.

**Comment by AdelAlm** [comment]
Here is my code:

```bash
#include <ArduinoBearSSL.h>
#include <ArduinoMqttClient.h>
#include <WiFi.h>

const char ssid[] = "XXXXXXXX";
const char pass[] = "YYYYYYYY";

WiFiClient    wifiClient;
BearSSLClient sslClient(wifiClient);
MqttClient    mqttClient(sslClient);

void setup() {
  Serial.begin(115200);
  while (!Serial);
}

void loop() {
  if (WiFi.status() != WL_CONNECTED) {
    connectWiFi();
  }
}

void connectWiFi() {
  Serial.print("Attempting to connect to SSID: ");
  Serial.print(ssid);
  Serial.print(" ");

  while (WiFi.begin(ssid, pass) != WL_CONNECTED) {
    Serial.print(".");
    delay(5000);
  }
  Serial.println();

  Serial.println("You're connected to the network");
  Serial.println();
}
```

It is a very light version of [Securely Connecting an Arduino MKR WiFi 1010 to AWS IoT Core](https://docs.arduino.cc/tutorials/mkr-wifi-1010/securely-connecting-an-arduino-mkr-wifi-1010-to-aws-iot-core/).

From my side, I get the same error when compiling with "Raspberry Pi Pico W" as board.

**Comment by earlephilhower** [comment]
It's basically trying to replace the BearSSL that's included with the core w/a much older version which doesn't have that function call.

You can hack out that call (assuming you don't care about MFLN) or move the ArduinoBearSSL library up to the latest release (which is close to 2 years old IIRC).
